### PR TITLE
refactor horovod.spark.run()

### PR DESCRIPTION
**Pull request by EnricoMi**
- makes horovod.spark.run() use horovod.run.mpi_run() and exposes extra_mpi_args
- adds lots of tests

Please do not merge into master, only code review it.

### CODE REVIEW

File: horovod/run/mpi_run.py [hunk 1]
```diff
@@ -88,7 +107,7 @@ def mpi_run(settings, common_intfs, env, command):
         common_intfs=','.join(common_intfs)) if common_intfs else ''
 
     # On large cluster runs (e.g. Summit), we need extra settings to work around OpenMPI issues
-    if settings.num_hosts >= 64:
+    if settings.num_hosts and settings.num_hosts >= _LARGE_CLUSTER_THRESHOLD:
         mpi_impl_flags.append('-mca plm_rsh_no_tree_spawn true')
```
Comment (alasdairm-gr) [inline horovod/run/mpi_run.py]: It's interesting that they introduced this change for mpi_run but not for the spark_run.  One of the reasons we override the mpi settings is to set this value to true.

File: horovod/spark/__init__.py [hunk 2]
```diff
@@ -193,28 +202,18 @@ def run(fn, args=(), kwargs={}, num_proc=None, start_timeout=None, env=None,
...
-                        env=' '.join('-x %s' % key for key in env.keys() if env_util.is_exportable(key)),
-                        python=sys.executable,
-                        encoded_driver_addresses=codec.dumps_base64(driver.addresses()),
-                        settings=codec.dumps_base64(settings)))
-        if settings.verbose >= 2:
-            print('+ %s' % mpirun_command)
-        exit_code = safe_shell_exec.execute(mpirun_command, env, stdout, stderr)
-        if exit_code != 0:
-            raise Exception('mpirun exited with code %d, see the error above.' % exit_code)
+        rsh_agent = (sys.executable,
```
Comment (alasdairm-gr) [inline horovod/spark/__init__.py]: Finally this code is easy to read :)

**Comment by alasdairm-gr** [comment]
It looks great.  This is far cleaner than what was there before.

File: horovod/run/mpi_run.py (hunk 1 above)
Comment (EnricoMi) [reply horovod/run/mpi_run.py]: horovod.spark.run does currently not set settings.num_hosts so there will be no way to reuse this code through horovod.spark.run. Maybe we could derive num_hosts from settings.hosts, then this extra args would be set automatically by mpi_run when needed.

**Comment by EnricoMi** [comment]
merged into upstream master

**Pull request closed without merging**